File: 63-unique-paths-ii/unique-paths-ii.java
class Solution {
    public int uniquePathsWithObstacles(int[][] obstacleGrid) {
        int m = obstacleGrid.length;
        int n = obstacleGrid[0].length;
        int[] prev = new int[n];
        int[] curr = new int[n];

        for(int i = 0 ; i < m ; i++){
            for(int j = 0 ; j < n ; j++){
                if(obstacleGrid[i][j] == 1){
                    curr[j] = 0;
                    continue;
                }
                if(i == 0 && j == 0){
                    curr[0] = 1;
                    continue;
                }

                int up = 0;
                int left = 0;
                if(i > 0){
                    up = prev[j];
                }

                if(j > 0){
                    left = curr[j-1];
                }

                curr[j] = up + left;
            }

            prev = curr;
        }
        return curr[n-1];
    }

   
}

File: 63-unique-paths-ii/unique-paths-ii.py
class Solution:
    def uniquePathsWithObstacles(self, grid: List[List[int]]) -> int:
        m = len(grid)
        n = len(grid[0])
        dp = [[0 for _ in range(n)] for _ in range(m)]

        if grid[0][0] == 1 or grid[m-1][n-1] == 1:
            return 0
        
        dp[m-1][n-1] = 1

        for i in range(m-1, -1, -1):
            for j in range(n-1, -1 , -1):

                if grid[i][j] == 1:
                    continue
                
                if i+1 < m:
                    dp[i][j] += dp[i+1][j]

                if j+1 < n:
                    dp[i][j] += dp[i][j+1]
        
        return dp[0][0]
